src/lib/privilege.c

```diff
@@ -156,6 +163,83 @@ void singularity_priv_escalate(void) {
         ABORT(255);
     }
 
+    singularity_message(DEBUG, "Clearing supplementary GIDs.\n");
+    if ( setgroups(0, NULL) == -1 ) {
+        singularity_message(ERROR, "Unable to clear the supplementary group IDs: %s (errno=%d).\n", strerror(errno), errno);
+        ABORT(255);
+    }
+    uinfo.dropped_groups = 1;
+
+}
+
+// Note that we don't initialize the singularity user at the beginning of the program;
+// this is because the 'singularity' user feature may not be used and hence not on the
+// system.
+//
+// This helps with sites upgrading from older versions without this user.
+static void cache_singularity_user(void) {
+
+    if (sinfo.ready == 1) {
+        return;
+    }
+
+    if ( uinfo.ready != 1 ) {
+        singularity_message(ERROR, "User info is not available\n");
+        ABORT(255);
+    }
+
+    if ( uinfo.userns_ready == 1 ) {
+        singularity_message(DEBUG, "User namespaces enabled; setting singularity user as the invoking user.\n");
+        sinfo.uid = uinfo.uid;
+        sinfo.gid = uinfo.gid;
+        return;
+    }
+
+    singularity_config_rewind();
+    char *username = singularity_config_get_value_default("singularity user", "singularity");
+    struct passwd *pw = getpwnam(username);
+    if (pw == NULL) {
+        singularity_message(ERROR, "Unable to determine UID/GID for user %s: %s (errno=%d)", username, strerror(errno), errno);
+        ABORT(255);
+    }
+    sinfo.uid = pw->pw_uid;
+    sinfo.gid = pw->pw_gid;
+}
+
+void singularity_priv_escalate_singularity(void) {
```

Comment:
I know this has already been merged, but what do you think about giving singularity_priv_escalate() an input variable? Something like `singularity_priv_escalate(char *user)`. If `char *user` is NULL then we escalate to root, otherwise we escalate to the user specified by `char *user`.

@bbockelm @gmkurtzer